### PR TITLE
Falcon-2052 Process SLA monitoring

### PR DESCRIPTION
**Pull request by PraveenAdlakha**

### CODE REVIEW

File: common/src/main/java/org/apache/falcon/entity/ProcessHelper.java [hunk 1]
```diff
@@ -185,4 +187,36 @@ private static void validateProcessInstance(Process process, Date instanceTime,
...
+        return cluster != null ? getSLA(cluster, process) : null;
+    }
+
+    public static Sla getSLA(Cluster cluster, Process process) {
+        final Sla clusterSla = cluster.getSla();
+        if (clusterSla != null) {
+            return clusterSla;
+        }
+        final Sla processSla = process.getSla();
+        return processSla == null ? null : processSla;
```
Comment (pallavi-rao) [inline common/src/main/java/org/apache/falcon/entity/ProcessHelper.java]: The additional null check seems unnecessary. You could just return process.getSla().

File: common/src/main/java/org/apache/falcon/entity/ProcessHelper.java [hunk 2]
```diff
@@ -185,4 +187,36 @@ private static void validateProcessInstance(Process process, Date instanceTime,
...
+        org.apache.falcon.entity.v0.process.Cluster cluster = getCluster(process, clusterName);
+        if (cluster == null) {
+            throw new FalconException("Invalid cluster: " + clusterName + " for process: " + process.getName());
+        }
+        return cluster.getValidity();
+    }
+
+    public static Sla getSLA(String clusterName, Process process) {
+        Cluster cluster = getCluster(process, clusterName);
+        return cluster != null ? getSLA(cluster, process) : null;
```
Comment (pallavi-rao) [inline common/src/main/java/org/apache/falcon/entity/ProcessHelper.java]: We should throw an exception here too, saying the cluster is invalid.

File: common/src/main/java/org/apache/falcon/entity/ProcessHelper.java [hunk 3]
```diff
@@ -185,4 +187,36 @@ private static void validateProcessInstance(Process process, Date instanceTime,
...
+    public static Sla getSLA(Cluster cluster, Process process) {
+        final Sla clusterSla = cluster.getSla();
+        if (clusterSla != null) {
+            return clusterSla;
+        }
+        final Sla processSla = process.getSla();
+        return processSla == null ? null : processSla;
+    }
+
+    public static Date getProcessValidityStart(Process process, String clusterName) throws FalconException {
```
Comment (pallavi-rao) [inline common/src/main/java/org/apache/falcon/entity/ProcessHelper.java]: Don't think we should add methods for such fine-grained elements.. getClusterValidity().getStart() should do (of course, with a null check).

File: common/src/main/java/org/apache/falcon/persistence/EntitySLAAlertBean.java [hunk 4]
```diff
@@ -37,31 +37,44 @@
  * */
 @Entity
 @NamedQueries({
-@NamedQuery(name = PersistenceConstants.GET_FEED_ALERTS, query = "select OBJECT(a) from FeedSLAAlertBean a where a.feedName = :feedName"),
```
Comment (pallavi-rao) [inline common/src/main/java/org/apache/falcon/persistence/EntitySLAAlertBean.java]: Corresponding PersistenceConstants constants should also be renamed.

File: common/src/main/java/org/apache/falcon/persistence/EntitySLAAlertBean.java [hunk 5]
```diff
@@ -37,31 +37,44 @@
...
-    @NamedQuery(name = PersistenceConstants.UPDATE_SLA_HIGH, query = "update FeedSLAAlertBean a set a.isSLAHighMissed = true where a.feedName = :feedName and a.clusterName = :clusterName and a.nominalTime = :nominalTime"),
-@NamedQuery(name = PersistenceConstants.GET_FEED_ALERT_INSTANCE, query = "select OBJECT(a) from FeedSLAAlertBean a where a.feedName = :feedName and a.clusterName = :clusterName and a.nominalTime = :nominalTime "),
- @NamedQuery(name = PersistenceConstants.DELETE_FEED_ALERT_INSTANCE, query = "delete from FeedSLAAlertBean a where a.feedName = :feedName and a.clusterName = :clusterName and a.nominalTime = :nominalTime")
+@NamedQuery(name = PersistenceConstants.GET_SLA_HIGH_CANDIDATES, query = "select OBJECT(a) from EntitySLAAlertBean a where a.isSLALowMissed = true and a.isSLAHighMissed = false "),
+    @NamedQuery(name = PersistenceConstants.UPDATE_SLA_HIGH, query = "update EntitySLAAlertBean a set a.isSLAHighMissed = true where a.entityName = :entityName and a.clusterName = :clusterName and a.nominalTime = :nominalTime"),
```
Comment (pallavi-rao) [inline common/src/main/java/org/apache/falcon/persistence/EntitySLAAlertBean.java]: Add entityType

File: docs/src/site/twiki/restapi/FeedLookup.twiki [hunk 6]
```diff
@@ -24,12 +24,12 @@ GET http://localhost:15000/api/entities/lookup/feed?path=/data/project1/2014/10/
     "feeds":
     [
         {
-           "feedName": "My-Feed1",
```
Comment (pallavi-rao) [inline docs/src/site/twiki/restapi/FeedLookup.twiki]: How is this change related?

File: falcon-regression/merlin/src/test/resources/OozieExampleInputData/normalInput/dataFile2.txt [hunk 7]
```diff
@@ -98,8 +98,8 @@
 //		return this.msgs.get(ARG.FEED_NAME);
 //	}
 //
-//	public void setFeedName(String feedName) {
```
Comment (pallavi-rao) [inline falcon-regression/merlin/src/test/resources/OozieExampleInputData/normalInput/dataFile2.txt]: How is this change related?

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java [hunk 8]
```diff
@@ -42,23 +43,25 @@ private EntityManager getEntityManager() {
     }
 
 
-    public void putMonitoredFeed(String feedName){
+    public void putMonitoredFeed(String feedName, String entityType){
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: feedName -> entityName

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java [hunk 9]
```diff
@@ -42,23 +43,25 @@ private EntityManager getEntityManager() {
...
 
-        MonitoredFeedsBean monitoredFeedsBean = new MonitoredFeedsBean();
-        monitoredFeedsBean.setFeedName(feedName);
+        MonitoredEntityBean monitoredEntityBean = new MonitoredEntityBean();
+        monitoredEntityBean.setFeedName(feedName);
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: setEntityName ?

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java [hunk 10]
```diff
@@ -42,23 +43,25 @@ private EntityManager getEntityManager() {
...
             beginTransaction(entityManager);
-            entityManager.persist(monitoredFeedsBean);
+            entityManager.persist(monitoredEntityBean);
         } finally {
             commitAndCloseTransaction(entityManager);
         }
     }
 
-    public MonitoredFeedsBean getMonitoredFeed(String feedName){
+    public MonitoredEntityBean getMonitoredFeed(String entityName, String entityType){
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: getMonitoredEntity?

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java [hunk 11]
```diff
@@ -67,82 +70,88 @@ public MonitoredFeedsBean getMonitoredFeed(String feedName){
...
+    public List<MonitoredEntityBean> getAllMonitoredFeed() throws ResultNotFoundException {
         EntityManager entityManager = getEntityManager();
         Query q = entityManager.createNamedQuery(PersistenceConstants.GET_ALL_MONITORING_FEEDS);
         List result = q.getResultList();
         entityManager.close();
         return result;
     }
 
-    public Date getLastInstanceTime(String feedName) throws ResultNotFoundException {
+    public Date getLastInstanceTime(String feedName , String entityType) throws ResultNotFoundException {
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: feedName -> entityName

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java [hunk 12]
```diff
@@ -67,82 +70,88 @@ public MonitoredFeedsBean getMonitoredFeed(String feedName){
...
-        q.setParameter("feedName", feedName);
+        q.setParameter("entityName", feedName);
+        q.setParameter("entityType", entityType);
         Date result = (Date)q.getSingleResult();
         entityManager.close();
         return result;
     }
 
-    public void deletePendingInstance(String feedName, String clusterName , Date nominalTime){
+    public void deletePendingInstance(String feedName, String clusterName , Date nominalTime, String entityType){
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: feedName -> entityName

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java [hunk 13]
```diff
@@ -67,82 +70,88 @@ public MonitoredFeedsBean getMonitoredFeed(String feedName){
...
+        q.setParameter("entityType", entityType);
         try{
             q.executeUpdate();
         } finally {
             commitAndCloseTransaction(entityManager);
         }
     }
 
-    public void deletePendingInstances(String feedName, String clusterName){
+    public void deletePendingInstances(String feedName, String clusterName, String entityType){
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: feedName -> entityName

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java [hunk 14]
```diff
@@ -67,82 +70,88 @@ public MonitoredFeedsBean getMonitoredFeed(String feedName){
...
+        q.setParameter("entityType", entityType);
         try{
             q.executeUpdate();
         } finally {
             commitAndCloseTransaction(entityManager);
         }
     }
 
-    public void putPendingInstances(String feed, String clusterName, Date nominalTime){
+    public void putPendingInstances(String feed, String clusterName, Date nominalTime, String entityType){
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: feedName -> entityName

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java [hunk 15]
```diff
@@ -168,7 +177,8 @@ private void commitAndCloseTransaction(EntityManager entityManager) {
         entityManager.close();
     }
 
-    public PendingInstanceBean getPendingInstance(String feedName, String clusterName, Date nominalTime) {
+    public PendingInstanceBean getPendingInstance(String feedName, String clusterName, Date nominalTime,
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: feedName -> entityName

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java [hunk 16]
```diff
@@ -177,40 +187,44 @@ public PendingInstanceBean getPendingInstance(String feedName, String clusterNam
...
+        q.setParameter("entityType", entityType);
         try {
             return q.getSingleResult();
         } finally {
             commitAndCloseTransaction(entityManager);
         }
     }
 
-    public FeedSLAAlertBean getFeedAlertInstance(String feedName, String clusterName, Date nominalTime) {
+    public EntitySLAAlertBean getFeedAlertInstance(String feedName, String clusterName, Date nominalTime,
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: feedName -> entityName, getFeedAlertInstance -> getEntityAlertInstance

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java [hunk 17]
```diff
@@ -177,40 +187,44 @@ public PendingInstanceBean getPendingInstance(String feedName, String clusterNam
...
         try {
             return q.getSingleResult();
         } finally {
             commitAndCloseTransaction(entityManager);
         }
     }
 
-    public void putSLAAlertInstance(String feedName, String cluster, Date nominalTime, Boolean isSLALowMissed,
-                                    Boolean isSLAHighMissed) {
+    public void putSLAAlertInstance(String feedName, String cluster, String entityType, Date nominalTime,
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: feedName -> entityName

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java [hunk 18]
```diff
@@ -230,13 +244,14 @@ public void updateSLAAlertInstance(String feedName, String clusterName, Date nom
         }
     }
 
-    public void deleteFeedAlertInstance(String feedName, String clusterName, Date nominalTime){
+    public void deleteFeedAlertInstance(String feedName, String clusterName, Date nominalTime, String entityType){
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: feedName -> entityName

File: prism/src/main/java/org/apache/falcon/service/EntitySLAListener.java [hunk 19]
```diff
@@ -18,14 +18,13 @@
...
-import org.apache.falcon.entity.v0.EntityType;
 
 import java.util.Date;
 
 /**
  * Interface for FeedSLAAlert to be used by Listeners.
  */
 public interface EntitySLAListener {
-    void highSLAMissed(String enityName, EntityType entityType, String clusterName, Date nominalTime)
+    void highSLAMissed(String feedName, String clusterName, String entityType, Date nominalTime)
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAListener.java]: Why change from enum to String? Why has the variable been renamed?

File: prism/src/main/java/org/apache/falcon/service/FeedSLAAlertService.java [hunk 20]
```diff
@@ -27,11 +27,8 @@
 import java.util.ArrayList;
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/FeedSLAAlertService.java]: Class rename?

File: prism/src/main/java/org/apache/falcon/service/FeedSLAAlertService.java [hunk 21]
```diff
@@ -117,33 +114,37 @@ void processSLACandidates(){
...
                 SchedulableEntityInstance schedulableEntityInstance = schedulableEntityList.get(0);
 
 
                 if (schedulableEntityInstance.getTags().contains(FeedSLAMonitoringService.get().TAG_WARN)) {
-                    store.putSLAAlertInstance(feedName, clusterName, nominalTime, true, false);
+                    store.putSLAAlertInstance(entityName, clusterName, EntityType.FEED.toString(),
+                            nominalTime, true, false);
                     //Mark in DB as SLA missed
-                    LOG.info("Feed :"+ feedName
+                    LOG.info("Feed :"+ entityName
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/FeedSLAAlertService.java]: Feed -> Entity

File: prism/src/main/java/org/apache/falcon/service/FeedSLAAlertService.java [hunk 22]
```diff
@@ -117,33 +114,37 @@ void processSLACandidates(){
...
+                            entityType);
                     return;
                 }
                 List<SchedulableEntityInstance> schedulableEntityList = new ArrayList<>(schedulableEntityInstances);
                 SchedulableEntityInstance schedulableEntityInstance = schedulableEntityList.get(0);
 
 
                 if (schedulableEntityInstance.getTags().contains(FeedSLAMonitoringService.get().TAG_WARN)) {
-                    store.putSLAAlertInstance(feedName, clusterName, nominalTime, true, false);
+                    store.putSLAAlertInstance(entityName, clusterName, EntityType.FEED.toString(),
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/FeedSLAAlertService.java]: you can use entityType variable rather than hardcode FEED. I understand that this branch is possible for feed.

File: prism/src/main/java/org/apache/falcon/service/FeedSLAAlertService.java [hunk 23]
```diff
@@ -153,11 +154,11 @@ void processSLACandidates(){
     }
 
     @Override
-    public void highSLAMissed(String feedName, EntityType entityType, String clusterName, Date nominalTime)
-        throws FalconException{
+    public void highSLAMissed(String feedName, String clusterName, String entityType , Date nominalTime
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/FeedSLAAlertService.java]: feedName -> entityName

File: prism/src/main/java/org/apache/falcon/service/FeedSLAMonitoringService.java [hunk 24]
```diff
@@ -17,9 +17,10 @@
  */
 package org.apache.falcon.service;
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/FeedSLAMonitoringService.java]: Class rename?

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java [hunk 25]
```diff
@@ -67,82 +70,88 @@ public MonitoredFeedsBean getMonitoredFeed(String feedName){
         } finally {
             entityManager.close();
         }
-        return ((MonitoredFeedsBean)result.get(0));
+        return ((MonitoredEntityBean)result.get(0));
     }
 
-    public void deleteMonitoringFeed(String feedName) {
+    public void deleteMonitoringFeed(String entityName, String entityType) {
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: In this and all other methods, it is better to use EntityType enum in the method signatures.

File: client/src/main/resources/process-0.1.xsd [hunk 26]
```diff
@@ -218,6 +217,7 @@
         </xs:annotation>
         <xs:sequence>
             <xs:element type="validity" name="validity"/>
+            <xs:element type="sla" name="sla" minOccurs="0" maxOccurs="1"/>
```
Comment (pallavi-rao) [inline client/src/main/resources/process-0.1.xsd]: Can a user specify sla at the cluster level only and not specify at the process level? The current xsd allows it. If yes, the parser has to be updated accordingly and also the MonitoringService class. If no, the parser should throw an error.

**Comment by pallavi-rao** [comment]
@PraveenAdlakha, I did half the review… but, there are too many inconsistencies with renaming. It gets a little confusing. Can you please fix the renames before I proceed further?

File: common/src/main/java/org/apache/falcon/entity/ProcessHelper.java (hunk 3 above)
Comment (PraveenAdlakha) [inline common/src/main/java/org/apache/falcon/entity/ProcessHelper.java]: in this method we just checking if the cluster that we have send to the function exits in the process definition...if we remove the method we have to do that check there.
Let me know your thought ? We have same kind of implementation for feed also.

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java (hunk 25 above)
Comment (PraveenAdlakha) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: agreed making the changes.

File: prism/src/main/java/org/apache/falcon/service/EntitySLAListener.java (hunk 19 above)
Comment (PraveenAdlakha) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAListener.java]: Same reason as I have string information available after reading the DB and I have to create entityType Object to send it to listeners.I thought they can create it at there end if needed.

File: prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java (hunk 25 above)
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/jdbc/MonitoringJdbcStateStore.java]: To pass the entity type to the query you just need to do entityType.name() and to convert to EntityType you need to do EntityType.valueOf(). This will ensure that invalid values are not passed.

File: common/src/main/java/org/apache/falcon/entity/ProcessHelper.java (hunk 3 above)
Comment (pallavi-rao) [inline common/src/main/java/org/apache/falcon/entity/ProcessHelper.java]: If this method is used in multiple places, guess it is fine, as the code doesn't get repeated.

File: client/src/main/resources/feed-0.1.xsd [hunk 27]
```diff
@@ -219,7 +219,7 @@
     <xs:complexType name="sla">
         <xs:annotation>
             <xs:documentation>
-                sla has two properties - slaLow and slaHigh. Both slaLow and slaHigh
+                sla has two properties - slaLow and slaHigh. Both PRoLow and slaHigh
```
Comment (pallavi-rao) [inline client/src/main/resources/feed-0.1.xsd]: accidental change?

Comment (PraveenAdlakha) [inline client/src/main/resources/feed-0.1.xsd]: yes will change it.

File: common/src/main/java/org/apache/falcon/entity/ProcessHelper.java [hunk 28]
```diff
@@ -185,4 +187,38 @@ private static void validateProcessInstance(Process process, Date instanceTime,
...
+        return process.getSla();
+    }
+
+    public static Date getProcessValidityStart(Process process, String clusterName) throws FalconException {
+        Cluster processCluster = getCluster(process, clusterName);
+        if (processCluster != null) {
+            return processCluster.getValidity().getStart();
+        } else {
+            throw new FalconException("No matching cluster " + clusterName
+                    + "found for feed " + process.getName());
```
Comment (pallavi-rao) [inline common/src/main/java/org/apache/falcon/entity/ProcessHelper.java]: process

File: prism/src/main/java/org/apache/falcon/resource/AbstractSchedulableEntityManager.java [hunk 29]
```diff
@@ -162,11 +162,11 @@ public SchedulableEntityInstanceResult getFeedSLAMissPendingAlerts(String feedNa
             Date end = (endStr == null) ? new Date() : EntityUtil.parseDateUTC(endStr);
 
             if (StringUtils.isBlank(feedName)) {
-                instances.addAll(FeedSLAMonitoringService.get().getFeedSLAMissPendingAlerts(start, end));
+                instances.addAll(EntitySLAMonitoringService.get().getFeedSLAMissPendingAlerts(start, end));
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/resource/AbstractSchedulableEntityManager.java]: These changes are fine, but, since there is an API exposed for the FeedSLA, we'll need to extend that API and add/modify corresponding methods to return process SLA misses.

File: prism/src/main/java/org/apache/falcon/service/EntitySLAListener.java [hunk 30]
```diff
@@ -18,14 +18,13 @@
...
-import org.apache.falcon.entity.v0.EntityType;
 
 import java.util.Date;
 
 /**
  * Interface for FeedSLAAlert to be used by Listeners.
  */
 public interface EntitySLAListener {
-    void highSLAMissed(String enityName, EntityType entityType, String clusterName, Date nominalTime)
+    void highSLAMissed(String entityName, String clusterName, String entityType, Date nominalTime)
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAListener.java]: I understand your need to keep entityType a string in the beans. But, the interface should use the enum.

File: prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java [hunk 31]
```diff
@@ -116,34 +125,71 @@ public static FeedSLAMonitoringService get() {
...
             Feed feed = (Feed) entity;
             // currently sla service is enabled only for fileSystemStorage
             if (feed.getLocations() != null) {
-                Set<String> currentClusters = DeploymentUtil.getCurrentClusters();
                 for (Cluster cluster : feed.getClusters().getClusters()) {
                     if (currentClusters.contains(cluster.getName())) {
                         if (FeedHelper.getSLA(cluster, feed) != null) {
                             LOG.debug("Adding feed:{} for monitoring", feed.getName());
-                            MONITORING_JDBC_STATE_STORE.putMonitoredFeed(feed.getName());
+                            MONITORING_JDBC_STATE_STORE.putMonitoredFeed(feed.getName(), EntityType.FEED.toString());
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java]: putMonitoredEntity

File: prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java [hunk 32]
```diff
@@ -116,34 +125,71 @@ public static FeedSLAMonitoringService get() {
...
     @Override
     public void onAdd(Entity entity) throws FalconException {
+        Set<String> currentClusters = DeploymentUtil.getCurrentClusters();
         if (entity.getEntityType() == EntityType.FEED) {
             Feed feed = (Feed) entity;
             // currently sla service is enabled only for fileSystemStorage
             if (feed.getLocations() != null) {
-                Set<String> currentClusters = DeploymentUtil.getCurrentClusters();
                 for (Cluster cluster : feed.getClusters().getClusters()) {
                     if (currentClusters.contains(cluster.getName())) {
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java]: Adding one entry per cluster will cause multiple entries for the same feed in the table when a feed has multiple clusters.

File: prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java [hunk 33]
```diff
@@ -116,34 +125,71 @@ public static FeedSLAMonitoringService get() {
...
-                            MONITORING_JDBC_STATE_STORE.putMonitoredFeed(feed.getName());
+                            MONITORING_JDBC_STATE_STORE.putMonitoredFeed(feed.getName(), EntityType.FEED.toString());
                         }
                     }
                 }
             }
         }
+        if (entity.getEntityType() == EntityType.PROCESS){
+            Process process = (Process) entity;
+            if (process.getSla() != null || checkProcessClusterSLA(process)){
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java]: Same can add similar check for FEED too?

File: prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java [hunk 34]
```diff
@@ -164,6 +210,7 @@ private boolean isSLAMonitoringEnabledInCurrentColo(Feed feed) {
 
     @Override
     public void onChange(Entity oldEntity, Entity newEntity) throws FalconException {
+        //TODO add code for process as an entity after discussion
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java]: Guess we can knock this TODO off and code?

File: prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java [hunk 35]
```diff
@@ -194,7 +242,7 @@ public void onReload(Entity entity) throws FalconException {
 
     @Override
     public String getName() {
-        return FeedSLAMonitoringService.class.getSimpleName();
+        return EntitySLAMonitoringService.class.getSimpleName();
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java]: Will need to update startup.properties too.

File: prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java [hunk 36]
```diff
@@ -305,40 +370,57 @@ void addNewPendingFeedInstances(Date to) throws FalconException {
...
+                    MONITORING_JDBC_STATE_STORE.deletePendingInstance(pendingInstanceBean.getEntityName(),
+                            pendingInstanceBean.getClusterName(), date, EntityType.FEED.toString());
                 }
             }
         }
     }
 
     // checks whether a given feed instance is available or not
-    private boolean checkFeedInstanceAvailability(String feedName, String clusterName, Date nominalTime) throws
+    private boolean checkFeedInstanceAvailability(String entityName, String clusterName, Date nominalTime,
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java]: rename method

File: prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java [hunk 37]
```diff
@@ -383,35 +465,54 @@ private boolean checkFeedInstanceAvailability(String feedName, String clusterNam
...
+     * @param entityName name of the feed
      * @param clusterName cluster name
      * @param start start time, inclusive
      * @param end end time, inclusive
      * @return Pending feed instances of the given feed which belong to the given time range and have missed SLA.
      * @throws FalconException
      */
-    public Set<SchedulableEntityInstance> getFeedSLAMissPendingAlerts(String feedName, String clusterName,
-                                                                 Date start, Date end) throws FalconException {
+    public Set<SchedulableEntityInstance> getFeedSLAMissPendingAlerts(String entityName, String clusterName,
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java]: method rename

File: prism/src/main/java/org/apache/falcon/handler/SLAMonitoringHandler.java [hunk 38]
```diff
@@ -60,7 +60,8 @@ private void updateSLAMonitoring(String clusterName, String[] outputFeedNamesLis
                 String templatePath = new Path(storage.getUriTemplate(LocationType.DATA)).toUri().getPath();
                 Date date = FeedHelper.getDate(templatePath, new Path(outputFeedInstancePathsList[index]),
                     EntityUtil.getTimeZone(feed));
-                FeedSLAMonitoringService.get().makeFeedInstanceAvailable(outputFeedNamesList[index], clusterName, date);
+                EntitySLAMonitoringService.get().makeFeedInstanceAvailable(outputFeedNamesList[index],
+                        clusterName, date);
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/handler/SLAMonitoringHandler.java]: Update this class to invoke makeProcessInstanceAvailable (you'll need to add this method) when a process completes successfully.

File: prism/src/main/java/org/apache/falcon/handler/SLAMonitoringHandler.java [hunk 39]
```diff
@@ -45,8 +45,14 @@
     @Override
     public void onSuccess(WorkflowExecutionContext context) throws FalconException {
         if (context.hasWorkflowSucceeded()) {
-            updateSLAMonitoring(context.getClusterName(), context.getOutputFeedNamesList(),
+            if (context.getEntityType().toString().equals(EntityType.FEED.toString())){
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/handler/SLAMonitoringHandler.java]: context.getEntityType().equals(EntityType.FEED.name()) should do.

File: prism/src/main/java/org/apache/falcon/handler/SLAMonitoringHandler.java [hunk 40]
```diff
@@ -45,8 +45,14 @@
...
+                updateSLAMonitoring(context.getClusterName(), context.getOutputFeedNamesList(),
                     context.getOutputFeedInstancePathsList());
+            }
+            if (context.getEntityType().toString().equals(EntityType.PROCESS.toString())){
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/handler/SLAMonitoringHandler.java]: context.getEntityType().equals(EntityType.FEED.name()) should do.

File: prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java [hunk 41]
```diff
@@ -579,4 +624,21 @@ Date getInitialStartTime(Entity entity, String clusterName, String entityType) t
...
+        Date nominalTime = null;
+        try {
+            nominalTime = DateUtil.parseDateFalconTZ(date);
+        }catch (ParseException e){
+            LOG.error("Exception while translating the date:", e);
+        }
+        if (nominalTime!= null){
+            List<Date> instances = (MONITORING_JDBC_STATE_STORE.getNominalInstances(entityName, clusterName,
+                entityType));
+            if (CollectionUtils.isEmpty(instances)){
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java]: should be ! empty?

File: prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java [hunk 42]
```diff
@@ -579,4 +624,21 @@ Date getInitialStartTime(Entity entity, String clusterName, String entityType) t
             return startTime.before(slaTime) ? startTime : slaTime;
         }
     }
+
+    public void updateProcessSLAMonitoring(String clusterName, String entityName, String date, String entityType) {
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/EntitySLAMonitoringService.java]: Better to have the method name synonymous to what is there for feed makeProcessInstanceAvailable.

File: prism/src/test/java/org/apache/falcon/service/EntitySLAAlertServiceTest.java [hunk 43]
```diff
@@ -136,25 +137,67 @@ public static void processSLALowCandidates() throws FalconException, Interrupted
...
+            ConfigurationStore.get().publish(EntityType.CLUSTER, processCluster);
+        }
+        org.apache.falcon.entity.v0.process.Sla sla = new org.apache.falcon.entity.v0.process.Sla();
+        Frequency processFrequency = new Frequency("1", Frequency.TimeUnit.minutes);
+        sla.setShouldEndIn(processFrequency);
+        process.setSla(sla);
+
+
+        Thread.sleep(10*1000);
+        System.out.println(monitoringJdbcStateStore.getEntityAlertInstance("test-process", "test-cluster", dateOne,
```
Comment (pallavi-rao) [inline prism/src/test/java/org/apache/falcon/service/EntitySLAAlertServiceTest.java]: No asserts in test case? Remove sysout.

File: prism/src/test/java/org/apache/falcon/jdbc/MonitoringJdbcStateStoreTest.java [hunk 44]
```diff
@@ -85,66 +86,72 @@ public void init() {
 
     @Test
     public void testInsertRetrieveAndUpdate() throws Exception {
```
Comment (pallavi-rao) [inline prism/src/test/java/org/apache/falcon/jdbc/MonitoringJdbcStateStoreTest.java]: Update the test to use some processes too, to ensure your changes are well tested.

**Comment by pallavi-rao** [comment]
+1, apart from the minor nit.

**Comment by ajayyadava** [comment]
:+1:  I think all review comments are addressed, will commit shortly.
